metadetect/lsst_measure.py

```diff
@@ -23,103 +22,236 @@
-        ormasks = _get_ormasks(sources=sources, exposure=exposure)
+        if use_deblended_stamps:
+            # remove all objects and replace with noise
+            noise_replacer = _get_noise_replacer(
+                exposure=exposure, sources=sources,
+            )
+        else:
+            noise_replacer = None
+
+        try:
+            results = _do_measure(
+                exposure=exposure,
+                sources=sources,
+                fitter=fitter,
+                stamp_size=stamp_size,
+                noise_replacer=noise_replacer,
+                meas_task=meas_task,
+                use_deblended_stamps=use_deblended_stamps,
+            )
+        finally:
+            if use_deblended_stamps:
+                # Insert all objects back into image
+                noise_replacer.end()
+    else:
+        results = []
+
+    return results
+
 
-        replacer = _get_noise_replacer(exposure=exposure, sources=sources)
+def _do_measure(
+    exposure,
+    sources,
+    fitter,
+    stamp_size,
+    noise_replacer=None,
+    meas_task=None,
+    use_deblended_stamps=DEFAULT_USE_DEBLENDED_STAMPS,
+):
+    """
+    run measurements on the input exposure, given the input measurement
+    task, list of sources, and fitter.  These steps are combined
+    because of the way that the deblending works to produce neighbor-subtracted
+    images, where the image is temporarily modified
+
+    Parameters
+    ----------
+    exp: ExposureF
+        The exposure on which to detect and measure
+    sources: list of sources
+        From a detection task
+    fitter: e.g. ngmix.gaussmom.GaussMom or ngmix.ksigmamom.KSigmaMom
+        For calculating moments
+    meas_task: SingleFrameMeasurementTask
+        An optional measurement task; if you already have centeroids etc. for
+        sources, no need to send it.  Otherwise this should do basic things
+        like finding the centroid
+    stamp_size: int
+        Size for postage stamps
+    use_deblended_stamps: bool
+        If True, use deblended postage stamps for the measurements.
+        Note deblending is run in either case
+    """
 
-        for i, source in enumerate(sources):
+    exp_bbox = exposure.getBBox()
+    results = []
+
+    # ormasks will be different within the loop below due to the replacer
+    ormasks = _get_ormasks(sources=sources, exposure=exposure)
+
+    for i, source in enumerate(sources):
 
-            # Skip parent objects where all children are inserted
-            if source.get('deblend_nChild') != 0:
-                continue
+        # Skip parent objects where all children are inserted
+        if source.get('deblend_nChild') != 0:
+            continue
 
-            ormask = ormasks[i]
+        ormask = ormasks[i]
 
+        if noise_replacer is not None:
             # This will insert a single source into the image
-            replacer.insertSource(source.getId())
+            noise_replacer.insertSource(source.getId())
```

Comment:
Now I am looking at these insertSource and removeSource calls and thinking we need another `try...finally` in here too.

Reply:
I thought the same, but getting this fixed up stream is probably not going to happen.

We could implement our own context managers

Reply:
A simpler option would be to make a copy of the exposure that we send to this code.

Reply:
In the `measure()` function we could replace the `try: finally:` with
```python
if use_deblended_stamps:
    exp_send = (copy of exposure)
else:
    exp_send = exposure
```
These coadds are small so this isn't much extra memory usage

Reply:
Are we worried about memory leaking from uncleaned up resources held in C++ even if we make copies?

Reply:
If an exception is thrown that must be caught, it will be an ngmix  exception, in which case we are OK.

Any exception from the lsst codes won't be caught and the code will crash and exit.

So I think we would be ok.

Reply:
```python
exp_send = afw_image.ExposureF(exposure, deep=True)
```

Reply:
OK. Let's remove the try-finally blocks and use a copy. Hopefully, that handles exceptions and won't corrupt data.